src/en/charms-upgrading.md

```diff
@@ -21,6 +21,21 @@ You can query for charm information, such as channels and revisions, with the
 `charm` utility, which is available via the Charm Tools software. See the
 [Charm Tools][charm-tools] page for guidance.
 
+## Upgrade mechanics
+
+Here is a summary of how charm upgrades work.
+
+The agent (running on each unit of the application being upgraded) unpacks the
+new version of the charm to a staging directory next to the original charm. It
+then points itself to the newly unpacked charm and deletes the old one, after
+which the expected hooks are fired using the newly unpacked charm code.
+
+The original charm files are not overwritten. The original charm continues to
```

Comment:
I don't understand this part. So, the new charm is in a completely new directory, but the old charm is still running in the original directory? They can't both be running at once? At some point the new charm has to take over and kill the process of the old one, surely?

Reply:
Just because there are two directories doesn't mean both their corresponding software is running simultaneously. Once the new one is unpacked, the original one stops running and the new one begins to run (hooks).

Reply:
I am not prepared to say what happens if the upgrade itself fails. I would like to hope that the old one will start to run again but I don't have that information.